### ISSUE TITLE
Null-scope Multitenant GW error when starting in client mode

### ISSUE THREAD

**Issue by Alfiva** [universAAL/remote#448]
_Originally Opened: @Alfiva (2014-11-17 11:27:12_)
_Originally Closed: 2015-07-27 13:51:54_

With multitenant GW 3.2.0 I have found the following exception when starting the GW in the &quot;client&quot; side :

      | 64 - mw.container.osgi - 3.2.0 | Session-&gt;unspecified method(): Failed t
o notify with success Gateway Singleton
java.lang.IllegalStateException: Scope cannot set be null, otherwise sending and
 rieving message will not work
        at org.universAAL.ri.gateway.Session.validateRemoteScope(Session.java:21
0)[98:ri.gateway.multitenant:3.2.0]
        at org.universAAL.ri.gateway.Session.send(Session.java:186)[98:ri.gatewa
y.multitenant:3.2.0]
        at org.universAAL.ri.gateway.Session$MessageSynchronizer.operate(Session
.java:102)[98:ri.gateway.multitenant:3.2.0]
        at org.universAAL.ri.gateway.Session$MessageSynchronizer.operate(Session
.java:96)[98:ri.gateway.multitenant:3.2.0]
        at org.universAAL.ri.gateway.utils.CallSynchronizer.performCall(CallSync
hronizer.java:79)[98:ri.gateway.multitenant:3.2.0]
        at org.universAAL.ri.gateway.Session.sendRequest(Session.java:251)[98:ri
.gateway.multitenant:3.2.0]

I am using the same configuration I had used when trying out the GW when it was in branch, 3.1.1-S version, so probably I am missing some configuration option to set a client scope (Nothing is mentioned in the wiki though):

remote-gateway-host=localhost
socket-port=8181
import-security-constraint-allow=.*
export-security-constraint-allow=.*
import-security-constraint-deny=none
export-security-constraint-deny=none
hash-key=temporary-key
connection-mode=client
routing-mode=forward

The connection is properly established because the command &quot;universaal:tenants&quot; shows the remote node, but the service calls from the client do not propagate (I get &quot;not succeeded&quot; in the client), as foretold by the error message.

--

From: _this issue has been automatically imported from our old issue tracker_

**Comment by Alfiva** [comment]
_Original author: @stefanolenzi-cnr (2014-11-18 17:37:15)_

Can you provide me the configuration that you are using for running the multiple instances of the GW?

**Comment by Alfiva** [comment]
_Original author: @Alfiva (2014-11-18 18:17:08)_

Two Karaf distros for windows, configured to get uAAL features 3.2.0, installed in both the uAAL-Minimal.Coordinator. Both with different AALSpace ID, peer role and slp port, both are COORDINATOR peers (they don't see each other).

Client GW configuration is shown in the first post. Server GW configuration is 

remote-gateway-socket=0.0.0.0
socket-port=8181
import-security-constraint-allow=.*
export-security-constraint-allow=.*
import-security-constraint-deny=none
export-security-constraint-deny=none
hash-key=temporary-key
connection-mode=server
routing-mode=router

I cant think of any other configuration that would affect this. Do you request something in particular?

**Comment by Alfiva** [comment]
_Original author: @Alfiva (2014-11-18 18:19:13)_

sorry, in my previous post I say they have different AND the same peer role (bad writing). They are both peer.role COORDINATOR.
And forgot to add that I am using the lightning example, so I have feature uAAL-Ont.Lightning in both sides, and the lighting server in the server and the lighting client in the client.

**Comment by Alfiva** [comment]
_Original author: @Alfiva (2014-11-20 11:05:04)_

Any news on this? I need this like for today... If it's not possible to solve this I'd like to change the code myself until it works, but that means I will not be using the release code in the demo, with the associated consequences.

**Comment by Alfiva** [comment]
_Original author: @Alfiva (2014-11-21 11:38:18)_

I found the problem, but it is pointless.

In ClientSocketCommunicationHandler.beforeRun() the session.setStatus(SessionEvent.SessionStatus.CONNECTED); is called BEFORE the session.setScope(SessionManager.getInstance()
                    .getAALSpaceIdFromSession(currentSession));

But setStatus already wants to send messages to synchronize with remote, but it can't cause the scope is not set yet. This is easily solved by calling setScope before setStatus (in two places actually).

However that is not the problem, as it still fails. I will open a new issue for this new bug, since this one is already identified and just waiting for code to be fixed.

**Comment by Alfiva** [comment]
_Original author: @Alfiva (2015-07-27 13:51:54)_

This is supposedly now fixed. I say supposedly because by fixing this new errors appeared: Because the setStatus calls now can take place, they encountered errors that didnt surface before because they were never called. These errors seem in all cases related to casting exceptions stemming from the security definition file. That will require its own bug fixing, but for the time being I added a &quot;feature&quot;: Security definition will no be checked if the corresponding property is not set in the config file of the GW.

As for the sentence &quot;However that is not the problem, as it still fails. I will open a new issue for this&quot;, this was because it missed the security definition file when I tried it.
